src/library/crate/cratestorage.cpp

```diff
@@ -0,0 +1,665 @@
+const QString CRATETABLE_LOCKED = "locked";
+
+const QString CRATE_SUMMARY_VIEW = "crate_summary";
+
+const QString CRATESUMMARY_TRACK_COUNT = "track_count";
+const QString CRATESUMMARY_TRACK_DURATION = "track_duration";
+
+const QString kCrateTracksJoin = QString(
+        "LEFT JOIN %3 ON %3.%4=%1.%2").arg(
+                CRATE_TABLE,
+                CRATETABLE_ID,
+                CRATE_TRACKS_TABLE,
+                CRATETRACKSTABLE_CRATEID);
+
+const QString kLibraryTracksJoin = kCrateTracksJoin + QString(
+        " LEFT JOIN %3 ON %3.%4=%1.%2").arg(
+                CRATE_TRACKS_TABLE,
+                CRATETRACKSTABLE_TRACKID,
+                LIBRARY_TABLE,
+                LIBRARYTABLE_ID);
+
+const QString kCrateSummaryViewSelect = QString(
+        "SELECT %1.*,"
+            "COUNT(CASE %2.%4 WHEN 0 THEN 1 ELSE NULL END) AS %5,"
+            "SUM(CASE %2.%4 WHEN 0 THEN %2.%3 ELSE 0 END) AS %6 "
+            "FROM %1").arg(
+                CRATE_TABLE,
+                LIBRARY_TABLE,
+                LIBRARYTABLE_DURATION,
+                LIBRARYTABLE_MIXXXDELETED,
+                CRATESUMMARY_TRACK_COUNT,
+                CRATESUMMARY_TRACK_DURATION);
+
+const QString kCrateSummaryViewQuery = QString(
+            "CREATE TEMPORARY VIEW IF NOT EXISTS %1 AS %2 %3 GROUP BY %4.%5").arg(
+                    CRATE_SUMMARY_VIEW,
+                    kCrateSummaryViewSelect,
+                    kLibraryTracksJoin,
+                    CRATE_TABLE,
+                    CRATETABLE_ID);
+
+
+class CrateQueryBinder {
+  public:
+    explicit CrateQueryBinder(FwdSqlQuery& query)
+        : m_query(query) {
+    }
+    virtual ~CrateQueryBinder() = default;
+
+    void bindId(const QString& placeholder, const Crate& crate) const {
+        m_query.bindValue(placeholder, crate.getId());
+    }
+    void bindName(const QString& placeholder, const Crate& crate) const {
+        m_query.bindValue(placeholder, crate.getName());
+    }
+    void bindLocked(const QString& placeholder, const Crate& crate) const {
+        m_query.bindValue(placeholder, crate.isLocked());
+    }
+    void bindAutoDjSource(const QString& placeholder, const Crate& crate) const {
+        m_query.bindValue(placeholder, crate.isAutoDjSource());
+    }
+
+  protected:
+    FwdSqlQuery& m_query;
+};
+
+
+class CrateSummaryQueryBinder: public CrateQueryBinder {
+  public:
+    explicit CrateSummaryQueryBinder(FwdSqlQuery& query)
+        : CrateQueryBinder(query) {
+    }
+    ~CrateSummaryQueryBinder() override = default;
+
+    void bindTrackCount(const QString& placeholder, const CrateSummary& crateSummary) const {
+        m_query.bindValue(placeholder, crateSummary.getTrackCount());
+    }
+    void bindTrackDuration(const QString& placeholder, const CrateSummary& crateSummary) const {
+        m_query.bindValue(placeholder, crateSummary.getTrackDuration());
+    }
+};
+
+} // anonymous namespace
+
+
+CrateQueryFields::CrateQueryFields(const FwdSqlQuery& query)
```

Comment:
What is the reason not to store query, than the member calls cannot fail and there is no need for the query parameter there. 

Or could query FwdSqlQuery become the base class?